Use a better hash seed when doing string hashing in the compiler

jrose-apple opened the pull request:
- Remove explicit seeds from string hashes that are compiler-version-dependent already.

- Comment that we can't change the seed for swiftdoc files (@nkcsgexi).

- Switch to a "better" explicit seed for serialization lookup tables. This last one changed the order of SIL entity deserialization (as @adrian-prantl noticed in the original #14890), but we shouldn't have been relying on that anyway (@gottesmm). It's actually rather scary that that was matching source order at the time.

jrose-apple committed:
Remove explicit seeds from string hashes that are compiler-dependent

The one for enum raw value checking isn't serialized anywhere, so we
can use normal in-process hashing. The ones in the Clang importer's
lookup tables are serialized, but the lookup table is already only
valid for a particular compiler build anyway (it goes into a Clang
PCM, which has the full compiler version in its cache key).

jrose-apple committed:
[Serialization] Switch to a better hash seed for lookup tables

...fulfilling the promised audit from 0747d9a339. No intended
functionality change /other/ than the order of already-unsorted lists.
This affected a number of SIL tests that relied on deserialization
order matching the original source order; I have no idea why the old
hash logic would make that the case. If we think that's a valuable
property, we should serialize a list of functions in addition to the
iterable table. (Maybe just in SIB mode?)

jrose-apple commented:
@swift-ci Please test

adrian-prantl commented on include/swift/Serialization/ModuleFormat.h:
```diff
@@ -52,7 +52,13 @@ const uint16_t SWIFTMODULE_VERSION_MAJOR = 0;
+///
+/// This is the same as the default used by llvm::djbHash, just provided
+/// explicitly here to note that it's part of the format.
+const uint32_t SWIFTMODULE_HASH_SEED = 5381;
```
Out of curiosity: Is the fact that the format depends on the hash seed a bug?

jrose-apple replied:
No, it's an OnDiskHashTable, so the hash used is inherently part of the format.

adrian-prantl replied:
Makes sense!

adrian-prantl commented on include/swift/Serialization/ModuleFormat.h:
```diff
@@ -52,7 +52,13 @@ const uint16_t SWIFTMODULE_VERSION_MAJOR = 0;
 /// it just ensures a conflict if two people change the module format.
 /// Don't worry about adhering to the 80-column limit for this line.
-const uint16_t SWIFTMODULE_VERSION_MINOR = 509; // [dynamic_lifetime] sil flag
+const uint16_t SWIFTMODULE_VERSION_MINOR = 510; // better string hash seed
```
What's up with the end-of-line comment both on the LHS and the RHS of the patch? Seems like neither one belongs on this line?

adrian-prantl replied:
Ah.. it's the last change that causes us to bump the version?

jrose-apple replied:
Yeah, see the comment above.

adrian-prantl commented on lib/Serialization/ModuleFile.cpp:
```diff
@@ -388,8 +388,7 @@ class ModuleFile::DeclTableInfo {
     if (key.first == DeclBaseName::Kind::Normal) {
-      // FIXME: DJB seed=0, audit whether the default seed could be used.
-      return llvm::djbHash(key.second, 0);
+      return llvm::djbHash(key.second, SWIFTMODULE_HASH_SEED);
```
Perhaps factor this out into a ModuleFile::hashFunction(). I'm not sure whether this is much better, I'll leave it up to you.

jrose-apple replied:
Yeah, I considered that, but figured this was slightly more likely to keep people from doing the wrong thing in the future. I'm not 100% sure of that though.

jrose-apple commented:
@swift-ci Please test

jrose-apple committed:
[Serialization] The hash seed for DeclCommentTableInfo can't change

Unlike compiled modules, swiftdoc files are considered a stable
format, so we can't change how information is stored in them. If we
add any more string-hashed tables to swiftdoc files, we should
consider using a new hash seed for those.

jrose-apple commented:
Updated to make the swiftdoc hash seed more explicit.

@swift-ci Please test

jrose-apple commented:
Hmm.

@swift-ci Please clean test

jrose-apple commented:
Definitely concerning. What changed between now and last time? Maybe someone added another hash?

jrose-apple commented:
Oops, yes, they did.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test

nkcsgexi reviewed (approved):
Keeping the hash seed for swiftdoc files makes sense to me. Thank you for the heads-up!

jrose-apple commented:
@swift-ci Please smoke test
